src/MessageBus/Handling/HandlerInvokingMiddleware.php

```diff
@@ -0,0 +1,47 @@
+<?php declare(strict_types = 1);
+
+namespace Damejidlo\MessageBus\Handling;
+
+use Damejidlo\MessageBus\IBusMessage;
+use Damejidlo\MessageBus\IMessageBusMiddleware;
+use Damejidlo\MessageBus\Middleware\MiddlewareCallback;
+use Damejidlo\MessageBus\Middleware\MiddlewareContext;
+
+
+
+final class HandlerInvokingMiddleware implements IMessageBusMiddleware
+{
+
+	/**
+	 * @var IHandlerProvider
+	 */
+	private $handlerProvider;
+
+	/**
+	 * @var IHandlerInvoker
+	 */
+	private $handlerInvoker;
+
+
+
+	public function __construct(IHandlerProvider $handlerProvider, IHandlerInvoker $handlerInvoker)
+	{
+		$this->handlerProvider = $handlerProvider;
+		$this->handlerInvoker = $handlerInvoker;
+	}
+
+
+
+	/**
+	 * @inheritDoc
+	 */
+	public function handle(IBusMessage $message, MiddlewareContext $context, MiddlewareCallback $nextMiddlewareCallback)
+	{
+		$handlerType = HandlerType::extractFrom($context);
+
+		$handler = $this->handlerProvider->get($handlerType);
+
+		return $this->handlerInvoker->invoke($handler, $message);
```

Comment:
There are several reasons for interface and separate service, none of them strong:
* to be able to fake the invoker in tests
* to offer extension point for future change of invoking strategy (e.g. `__invoke` method), even if we decided we do not want this behaviour to be user-extensible at the moment
* static analysis - without the custom validation it will fail (validation and proper exception is IMHO better than failing hard and the necessity of phpstan exclusion)

So, yes, I could call `$handler->handle($message)` directly in middleware and make static analysis exclusion. But this is the way I designed the flow (resolve handlers -> provide handler -> invoke handler) and set up the middleware stack and it seems better to me to have the middleware not doing the heavy lifting.

Reply:
(leaving this open for discussion)

Reply:
As discussed in person - let's leave it as it is now ¯\_(ツ)_/¯